alternator/streams.cc

```diff
@@ -180,5 +180,255 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
+// to maximum now() - confidence_interval will get all records 
+// up until that limit, regardless of late writes.
+static constexpr auto confidence_interval = 30s;
+
+/**
+ * This implementation of CDC stream to dynamo shard mapping
+ * is simply a 1:1 mapping, id=shard. Simple, and fairly easy
+ * to do ranged queries later.
+ * 
+ * Downside is that we get a _lot_ of shards. And with actual
+ * generations, we'll get hubba-hubba loads of them. 
+ * 
+ * Improvement ideas: 
+ * 1.) Prune expired streams that cannot have records (check ttls)
+ * 2.) Group cdc streams into N per shards (by token range)
+ * 
+ * The latter however makes it impossible to do a simple 
+ * range query with limit (we need to do staggered querying, because
+ * of how GetRecords work).
+ * The results of a select from cdc log on more than one cdc stream,
+ * where timestamp < T will typically look like: 
+ * 
+ * <shard 0> <ts0> ...
+ * ....
+ * <shard 0> <tsN> ...
+ * <shard 1> <ts0p> ...
+ * ...
+ * <shard 1> <tsNp> ...
+ *
+ * Where tsN < timestamp and tsNp < timestamp, but range [ts0-tsN] and [ts0p-tsNp]
+ * overlap. 
+ * 
+ * Now, if we have limit=X, we might end on a row inside shard 0.
+ * For our next query, we would in the simple case just say "timestamp > ts3 and timestamp < T",
+ * but of course this would then miss rows in shard 1 and above. 
+ * 
+ * However if we instead set the conditions to 
+ * 
+ *   "(token(stream_id) = token(shard 0) and timestamp > ts3 and timestamp < T)
+ *      or (token(stream_id) > token(shard 0) and timestamp < T)"
+ * 
+ * we can get the correct data. Question is however how well this will perform, if at 
+ * all.
+ * 
+ * For now, go simple, but maybe consider if this latter approach would work
+ * and perform. 
+ * 
+ */
+
+future<executor::request_return_type> executor::describe_stream(client_state& client_state, service_permit permit, rjson::value request) {
+    _stats.api_operations.describe_stream++;
+
+    auto stream_arn = rjson::get<alternator::stream_arn>(request, "StreamArn");
+    auto shard_start = rjson::get_opt<shard_id>(request, "ExclusiveStartShardId");
+    auto limit = rjson::get_opt<int>(request, "Limit").value_or(100); // according to spec
+    auto ret = rjson::empty_object();
+    auto stream_desc = rjson::empty_object();
+
+    auto& db = _proxy.get_db().local();
+    auto& cf = db.find_column_family(stream_arn);
+    auto bs = cdc::get_base_table(_proxy.get_db().local(), *cf.schema());
+    auto& opts = bs->cdc_options();
+
+    rjson::set(stream_desc, "StreamStatus", rjson::from_string(opts.enabled() ? "ENABLED"s : "DISABLED"s));
+
+    stream_view_type type = stream_view_type::KEYS_ONLY;
+
+    if (opts.preimage() && opts.postimage()) {
+        type = stream_view_type::NEW_AND_OLD_IMAGES;
+    } else if (opts.preimage()) {
+        type = stream_view_type::OLD_IMAGE;
+    } else if (opts.postimage()) {
+        type = stream_view_type::NEW_IMAGE;
+    }
```

Comment:
Since I eventually want to marshall variables as well, and I don't want to repeat the above code, having _one_ single operator>> and operator<< is imho much safer.